### ISSUE TITLE
Improved Pre-req/Co-req scraping

### ISSUE THREAD

**Issue by resurl** [ubclaunchpad/life-at-ubc#33]
## Background Info/Issues
- Pre-requisite/Co-requsite information from scraper is inaccurate

Currently the scraper just uses regex to match course codes in the `<p>` containing the pre-reqs (see examples if unclear), but it's a naive approach because the resulting information is only a list of _potential_ course prereqs. fixing this would be really useful when we implement relations between courses

if you need any help understanding the scraper code, with puppeteer, or want some ideas on how to approach it, send me a DM on slack! i tried to document the code as best i could but it's a bit of a big boi at first.

## TODO
- ~~change `scraper.ts` so that it provides more information about a requisite course~~
- build a lexer for requisite text so that it can categorize each req accurately 

## Related Links
https://pptr.dev/

## Examples
![chem130](https://user-images.githubusercontent.com/23158004/97091999-029dd100-15f5-11eb-8f07-f1fd9c487cae.jpg)
the scraper currently lists all of these as pre-reqs, when we want richer information, something like
```typescript
// array containing pre-reqs in output.json, this replaces the "coReq" and "preReq" fields in the 
// current scraper version
// prereq_option1 is like a tag that would group the "one of"s together
"reqs": [
    {
      "course": "CHEM 110",
      "requirement": "prereq_option1"
    },
    {
      "course": "CHEM 111",
      "requirement": "prereq_option1"
    },
...
  ],
```
![apsc-prereq-ex](https://user-images.githubusercontent.com/23158004/97092145-552bbd00-15f6-11eb-870d-c84d25a8e318.jpg)
```typescript
"reqs": [
    {
      "course": "APSC 172",
      "requirement": "prereq"
    },
    {
      "course": "APSC 173",
      "requirement": "coreq"
    }
  ],
```
![cpsc314-ex](https://user-images.githubusercontent.com/23158004/97092146-565cea00-15f6-11eb-80a6-8876acd39ea2.jpg)
```typescript
"reqs": [
    {
      "course": "CPSC 221",
      "requirement": "prereq"
    },
    {
      "course": "MATH 200",
      "requirement": "prereq_option1"
    },
   ...,
   {
      "course": "MATH 152",
      "requirement": "prereq_option2"
    }
  ],
```
![cpsc317](https://user-images.githubusercontent.com/23158004/97092149-5826ad80-15f6-11eb-9268-645a779d8540.jpg)
```typescript
"reqs": [
    {
      "course": "CPSC 213",
      "requirement": "prereq"
    },
    {
      "course": "CPSC 221",
      "requirement": "prereq"
    }
  ],
```

**Comment by yehee** [comment]
Just a heads up (and you probably are already aware of this), seems like all co-reqs in `output.json` are empty.

**Comment by resurl** [comment]
@yehee yes thank you! i forgot to mention that. i didn't think we'd work with the pre-req/co-req stuff so soon when i worked on it, but i'll try to get a quick fix working soon so we can at least test that data

**Comment by yehee** [comment]
Continuing on with this [thread](https://github.com/ubclaunchpad/life-at-ubc/pull/48#issuecomment-723245612), I think the case with CPSC 340 is especially tricky to handle. How's others look like in general? Would it be possible to get away with handling edge cases if there aren't that many? 🤔

**Comment by resurl** [comment]
@yehee yeah so! i was actually thinking about how to implement this today and it seems like the simplest approach would be to build a syntax tree from the text for prereqs/coreqs and parse it according to that somehow. its kinda more complicated than i'd like so maybe im overthinking it? but reqs vary pretty wildly from course to course and there are a ton of different cases 😔 there are quite a few faculty of science/applied science courses that have similar req structure as 340

~~after the parsing, i think most of the edge cases come from the unspecific requisites like "Precalc 12", "1/2/300-level course from x", "3 credits from x", so im still trying to figure out how we want to handle that too~~

actually, i'm wondering now if storing the raw text for reqs might make more sense than storing the course codes with relations, but i'll bring it up in better detail during the meeting tomorrow

**Comment by yehee** [comment]
@resurreccionl wrt storing the plain text, I think it would make sense to do that along with what we already have, since the info about the list of pre-req courses are still needed to offer features like explicitly flagging a course when they haven't met a pre-reqs, etc. But again, that feature would definitely be a stretch goal, so I'm ok having it as a plain text for now and revisit the topic later!

**Comment by resurreccionl** [comment]
@yehee i'm unfortunately not familiar with postgres but i keep hearing it's hard to change relational database schemas once it's already been defined, so i'm wondering if adding another section field to store the plaintext going to be much of a problem ? will you have to change much in the database?
